report/metrics.go

```diff
@@ -109,47 +107,35 @@ func last(t1, t2 time.Time) time.Time {
 	return t2
 }
 
-// concat returns a new list formed by adding each element of acc to curr
-// acc or curr can be nil.
-func concat(acc []interface{}, curr ps.List) ps.List {
-	if curr == nil {
-		curr = ps.NewList()
-	}
-	for i := len(acc) - 1; i >= 0; i-- {
-		curr = curr.Cons(acc[i])
-	}
-	return curr
-}
-
 // Add returns a new Metric with (t, v) added to its Samples. Add is the only
 // valid way to grow a Metric.
+// TODO: join t and v into a Sample to avoid extra allocations?
+// TODO: This seems to be too elaborate, Add() only seems to be used to add ordered Samples.
+//       Replace this by a specialized version getting a slice of ordered Samples
+//       without duplicates?
 func (m Metric) Add(t time.Time, v float64) Metric {
 	// Find the first element which is before you element, and insert
 	// your new element in the list.  NB we want to dedupe entries with
 	// equal timestamps.
-	// This should be O(1) to insert a latest element, and O(n) in general.
-	curr, acc := m.Samples, make([]interface{}, 0, m.Len()+1)
-	for {
-		if curr == nil || curr.IsNil() {
+	samplesOut := make([]Sample, 0, len(m.Samples)+1)
+	var i int
+	// TODO: use binary search + copy() to improve performance
+	for i = 0; i < len(m.Samples); i++ {
+		if m.Samples[i].Timestamp.Equal(t) {
+			i++
 			break
 		}
-
-		currSample := curr.Head().(Sample)
-		if currSample.Timestamp.Equal(t) {
-			curr = curr.Tail()
+		if m.Samples[i].Timestamp.After(t) {
 			break
 		}
-		if currSample.Timestamp.Before(t) {
-			break
-		}
-
-		acc, curr = append(acc, curr.Head()), curr.Tail()
+		samplesOut = append(samplesOut, m.Samples[i])
+	}
+	samplesOut = append(samplesOut, Sample{t, v})
+	if i < len(m.Samples) {
+		samplesOut = append(samplesOut, m.Samples[i:]...)
```

Comment:
Since the array is already allocated, would `copy(samplesOut[i:], m.Samples[i:])` be faster?